feat: Remove all proctortrack references and add support for requires_escalation_email and show_review_rules configs

mlabeeb03 opened the pull request:
## Description

All references to the hardcoded 'proctortrack' string have been removed from the codebase in addition to the SHOW_REVIEW_RULES waffle flag. It was used to determine whether an escalation email is required and whether review rules should be shown. These decisions are now made based on the value of 'requires_escalation_email' and 'show_review_rules' config in the course's proctoring backend.

Removed translation commands, mock variables and user facing strings that contained 'proctortrack'.

Updated all test cases that were using proctortrack to use the new config variables.

Add a test case that verifies the functionality of the new `show_review_rules` config.

## Testing information

Test cms/djangoapps/contentstore/api/views/course_validation.py by making a get request and verifying the correct value of `requires_escalation_email`:
<img width="1087" height="578" alt="Screenshot 2025-10-31 at 3 14 05 PM" src="https://github.com/user-attachments/assets/c8df3981-25d4-4cce-adc8-9b06994379d8" />


Test cms/djangoapps/models/settings/course_metadata.py by making sure the the `Proctoring exam escalation contact` is not visible when `requires_escalation_email` is not set in the proctoring backend:
<img width="1088" height="381" alt="Screenshot 2025-10-31 at 3 37 40 PM" src="https://github.com/user-attachments/assets/34dda9b9-5faa-4609-8015-3465beffba24" />

Also test that if you do not provide the escalation email when `requires_escalation_email` is set, you should see an error message in API call:
<img width="1093" height="658" alt="Screenshot 2025-10-31 at 3 53 15 PM" src="https://github.com/user-attachments/assets/6ac9c048-2c34-4f04-a400-35bf810c3dd6" />
You can see this error on the main course page as well as while updating the `Proctoring provider` in the advanced course settings.





## Supporting information

More details available here: https://github.com/openedx/edx-platform/issues/36329

openedx-webhooks commented:
Thanks for the pull request, @mlabeeb03!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

mlabeeb03 commented:
@kdmccormick the initial draft PR is ready, just FYI.

mlabeeb03 committed:
feat: Remove all proctortrack references and add requires_escalation_email and show_review_rules configs

All references to the hardcoded 'proctortrack' string have been removed from the codebase in addition to the SHOW_REVIEW_RULES waffle flag. It was used to determince whether an escalation email is required and whether review rules should be shown. These decisions are now made based on the value of 'requires_escalation_email' and 'show_review_rules' config in the default proctoring backend.

Removed translation commands, mock variables and user facing strings that contained 'proctortrack'.

Updated all test cases that were using proctortrack to use the new config variables.

See this ticket for further information:
  https://github.com/openedx/edx-platform/issues/36329

kdmccormick commented:
Thanks @mlabeeb03 , I will review shortly.

mlabeeb03 committed:
feat: update api to include list of proctoring backend that require an escalation email

This change is a requirement for frontend-app-learning. It will no longer use hardcoded strings to check if an escalation email is required or not. Instead the api will return the list of proctoring backends that require and escalation email.

mlabeeb03 commented:
Hey @kdmccormick I added a new commit. Updated the API because `frontend-app-authoring` now requires a list of proctoring backends that have `requires_escalation_email` set.

kdmccormick commented on xmodule/course_block.py:
```diff
@@ -296,6 +296,27 @@ def get_available_providers():  # lint-amnesty, pylint: disable=missing-function
     return available_providers
 
 
+def get_requires_escalation_email_providers():  # lint-amnesty, pylint: disable=missing-function-docstring
```
I know that the function above this one is missing a docstring, but that doesn't mean we should follow that bad pattern with new code. Could you add a docstring?

kdmccormick commented on cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py:
```diff
@@ -1323,7 +1316,7 @@ def create_xblock_info(  # lint-amnesty, pylint: disable=too-many-statements
                         "proctoring_exam_configuration_link": proctoring_exam_configuration_link,
                         "supports_onboarding": supports_onboarding,
-                        "show_review_rules": show_review_rules,
+                        "show_review_rules": show_review_rules(),
```
Rather than looking at the default proctoring backend's config, this value should be based on the config of `course.proctoring_provider`, which is not guaranteed to be the same as the DEFAULT.

kdmccormick commented on xmodule/course_block.py:
```diff
@@ -296,6 +296,27 @@ def get_available_providers():  # lint-amnesty, pylint: disable=missing-function
+        and proctoring_backend_settings[provider].get(
+            "requires_escalation_email", False
+        )
+    ]
```
openedx/envs/common.py defines PROCTORING_BACKENDS, so we can safely assume that it is always present on the `settings` object. No need to use `getattr`, which is error-prone and indirect.

```suggestion
    requires_escalation_email_providers = [
        provider
        for provider in settings.PROCTORING_BACKENDS
        if provider != "DEFAULT"
        and settings.PROCTORING_BACKENDS[provider].get(
            "requires_escalation_email", False
        )
    ]
```

kdmccormick commented on lms/djangoapps/instructor/views/instructor_dashboard.py:
```diff
@@ -312,7 +313,7 @@ def _section_special_exams(course, access):
         # Only call does_backend_support_onboarding if not using an LTI proctoring provider
         show_onboarding = does_backend_support_onboarding(course.proctoring_provider)
-        if proctoring_provider == 'proctortrack':
+        if requires_escalation_email():
```
Rather than looking at the default proctoring backend's config, this value should be based on the config of the course's `proctoring_provider`, which is not guaranteed to be the same as the DEFAULT.

kdmccormick commented on cms/djangoapps/contentstore/config/waffle.py:
```diff
@@ -22,10 +22,6 @@
     f'{WAFFLE_NAMESPACE}.enable_checklists_quality', __name__, LOG_PREFIX
 )
 
-SHOW_REVIEW_RULES_FLAG = CourseWaffleFlag(  # lint-amnesty, pylint: disable=toggle-missing-annotation
-    f'{WAFFLE_NAMESPACE}.show_review_rules', __name__, LOG_PREFIX
-)
-
```
With SHOW_REVIEW_RULES_FLAG  removed, can you ensure that there's a unit test somewhere which validates that the new `show_review_rules` option works as intended?

kdmccormick commented on lms/envs/common.py:
```diff
@@ -3340,14 +3340,6 @@
     'ROOT_PATH': '',
 }
 
-### Proctoring configuration (redirct URLs and keys shared between systems) ####
-PROCTORING_BACKENDS = {
-    'DEFAULT': 'null',
-    # The null key needs to be quoted because
-    # null is a language independent type in YAML
-    'null': {}
-}
-
```
Thanks for removing this. Can you also ensure that the PROCTORING_BACKENDS setting in openedx/envs/common.py is annotated with a `.. setting_name` and `.. setting_description`, which describes the shape of the dictionary and the meanings of the show_review_rules and requires_escalation_email keys?

kdmccormick reviewed (changes requested):
Thanks for the work so far @mlabeeb03

mlabeeb03 committed:
feat: update proctoring helper functions, add test case and refactoring

Co-authored-by: Kyle McCormick <kyle@kylemccormick.me>

mlabeeb03 commented:
@kdmccormick I've fixed the unit tests. You may review the PR.

kdmccormick commented on openedx/envs/common.py:
```diff
@@ -2402,7 +2402,31 @@ def should_send_learning_badge_events(settings):
+#             "<provider_name>": {
+#                 "show_review_rules": <bool>,
+#                 "requires_escalation_email": <bool>,
+#                 ... additional provider-specific options ...
+#             },
+#             ...
```
let's add these lines in order to explain how DEFAULT works
```suggestion
#             },
#             "<another_provider_name>": { ... }
#             ...
#             "DEFAULT": "<provider_name>",
```

kdmccormick commented on openedx/envs/common.py:
```diff
@@ -2402,7 +2402,31 @@ def should_send_learning_badge_events(settings):
+#
+#     **requires_escalation_email** (bool):
+#         Providers with this flag set to True must specify an escalation email address
+#         in the advanced course settings. Default is False.
```
```suggestion
#         Providers with this flag set to True require that an escalation email address be
#         specified in the advanced course settings. Default is False.
```
here's a minor rewording, clarifying that course teams specify an email address, not providers.

kdmccormick commented on common/djangoapps/util/proctoring.py:
```diff
@@ -0,0 +1,30 @@
+    Returns the value of 'requires_escalation_email' in the given proctoring backend.
+    The default value for 'requires_escalation_email' is False.
+    """
+    return get_proctoring_config("requires_escalation_email", False, proctoring_provider)
```
The way this function is coded now, I don't think `"DEFAULT"` will work a the default argument. The proctoring backends dictionary looks like this:
```
{
   "backend1": { "requires_escalation_email": True, ... },
   "backend2": {"requires_escalation_email": False,  ... },
   "DEFAULT: "backend1",
}
```
so if I look up the key `"DEFAULT"`, it does not return a config dict. It just returns a provider name (`"backend1"`), and then calling `"backend1".get("requires_escalation_email")` will raise an AttributeError. Right?

Rather than fix the behavior, I'd prefer you simplify this function. We don't need `"DEFAULT"` behavior, and we don't need `get_proctoring_config` to exist.  The implementation can be simply `settings.PROCTORING_BACKENDS.get(proctoring_provider, {}).get("requires_escalation_email", False)`. This is easier to read quickly and is less likely to have bugs hiding in it.

I have the same feedback for `show_review_rules`.

mlabeeb03 replied:
Nice catch, fixed it.

mlabeeb03 committed:
feat: simplify proctoring util functions and update comments

kdmccormick commented on xmodule/course_block.py:
```diff
@@ -296,6 +297,22 @@ def get_available_providers():  # lint-amnesty, pylint: disable=missing-function
 
 
+def get_requires_escalation_email_providers():
+    """Return list of available proctoring providers that require an escalation email."""
```
```suggestion
def get_requires_escalation_email_providers() -> list[str]:
    """
    Return list of available proctoring providers that require an escalation email.
    """
```
formatting & type annotation

kdmccormick replied:
@mlabeeb03 if you can just apply this suggestion and the other one, then I'll go ahead and merge

kdmccormick commented on xmodule/course_block.py:
```diff
@@ -283,7 +283,8 @@ def default(self):
 
-def get_available_providers():  # lint-amnesty, pylint: disable=missing-function-docstring
+def get_available_providers():
+    """Return list of available proctoring providers."""
```
```suggestion
def get_available_providers() -> list[str]:
    """
    Return list of available proctoring providers.
    """
```

kdmccormick reviewed (approved):
looks great, thanks!

mlabeeb03 commented:
@kdmccormick you can merge it now.
